File: yc266/1234.py
from sys import stdin


class SegmentTree:
    DEFAULT_VALUE = 10 ** 18
    DEFAULT_LAZY = 0
    op = min

    def __init__(self, size):
        self._size = size
        t = 1
        while t < size:
            t *= 2
        self._offset = t - 1
        self._values = [SegmentTree.DEFAULT_VALUE] * (t * 2 - 1)
        self._lazy = [SegmentTree.DEFAULT_LAZY] * (t * 2 - 1)

    def build(self, iterable):
        op = SegmentTree.op
        values = self._values
        values[self._offset:self._offset + self._size] = iterable
        for i in range(self._offset - 1, -1, -1):
            values[i] = SegmentTree.op(values[i * 2 + 1], values[i * 2 + 2])

    def _get_segmen_indexes(self, start, stop):
        ls = []
        rs = []
        l = start + self._offset
        r = stop + self._offset
        while l < r:
            if l & 1 == 0:
                ls.append(l)
            if r & 1 == 0:
                rs.append(r - 1)
            l = l // 2
            r = (r - 1) // 2
        return ls + rs[::-1]

    def _propagate(self, start, stop):
        lazy = self._lazy
        values = self._values
        indexes = []
        for i in self._get_segmen_indexes(start, stop):
            if i == 0:
                return
            while i != 0:
                i = (i - 1) // 2
                indexes.append(i)
        for i in sorted(set(indexes)):
            l = lazy[i]
            if l == SegmentTree.DEFAULT_LAZY:
                continue
            lazy[i * 2 + 1] += l
            values[i * 2 + 1] += l
            lazy[i * 2 + 2] += l
            values[i * 2 + 2] += l
            lazy[i] = SegmentTree.DEFAULT_LAZY

    def apply(self, start, stop, value):
        op = SegmentTree.op
        values = self._values
        lazy = self._lazy
        self._propagate(start, stop)
        for i in self._get_segmen_indexes(start, stop):
            lazy[i] += value
            values[i] += value
            while i >= 1:
                i = (i - 1) // 2
                values[i] = op(values[i * 2 + 1], values[i * 2 + 2])

    def query(self, start, stop):
        op = SegmentTree.op
        values = self._values
        self._propagate(start, stop)
        result = SegmentTree.DEFAULT_VALUE
        for i in self._get_segmen_indexes(start, stop):
            result = op(result, values[i])
        return result


readline = stdin.readline

N = int(readline())
a = list(map(int, readline().split()))
Q = int(readline())

st = SegmentTree(N)
st.build(a)
result = []
for _ in range(Q):
    k, l, r, c = map(int, readline().split())
    if k == 1:
        st.apply(l - 1, r, c)
    elif k == 2:
        result.append(st.query(l - 1, r))
print(*result, sep='\n')

File: yc266/1234.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

const (
	defaultValue = math.MaxInt64
	defaultLazy  = 0
)

type segmentTree struct {
	offset int
	values []int
	lazy   []int
}

func newSegmentTree(n int) segmentTree {
	var result segmentTree
	t := 1
	for t < n {
		t *= 2
	}
	result.offset = t - 1
	result.values = make([]int, 2*t-1)
	result.lazy = make([]int, 2*t-1)
	for i := 0; i < 2*t-1; i++ {
		result.values[i] = defaultValue
		result.lazy[i] = defaultLazy
	}
	return result
}

func op(x, y int) int {
	return min(x, y)
}

func (st segmentTree) build(a []int) {
	for i, v := range a {
		st.values[st.offset+i] = v
	}
	for i := st.offset - 1; i > -1; i-- {
		st.values[i] = op(st.values[i*2+1], st.values[i*2+2])
	}
}

func (st segmentTree) segments(start, stop int) []int {
	result := make([]int, 0, 20)
	l := start + st.offset
	r := stop + st.offset
	for l < r {
		if l&1 == 0 {
			result = append(result, l)
		}
		if r&1 == 0 {
			result = append(result, r-1)
		}
		l = l / 2
		r = (r - 1) / 2
	}
	return result
}

func (st segmentTree) propagate(segments []int) {
	for _, i := range segments {
		indexes := make([]int, 0, 20)
		for i != 0 {
			i = (i - 1) / 2
			indexes = append(indexes, i)
		}
		for len(indexes) != 0 {
			j := indexes[len(indexes)-1]
			indexes = indexes[:len(indexes)-1]
			if st.lazy[j] == defaultLazy {
				continue
			}
			st.lazy[j*2+1] += st.lazy[j]
			st.values[j*2+1] += st.lazy[j]
			st.lazy[j*2+2] += st.lazy[j]
			st.values[j*2+2] += st.lazy[j]
			st.lazy[j] = defaultLazy
		}
	}
}

func (st segmentTree) apply(start, stop int, value int) {
	segments := st.segments(start, stop)
	st.propagate(segments)
	for _, i := range segments {
		st.lazy[i] += value
		st.values[i] += value
		for i != 0 {
			i = (i - 1) / 2
			st.values[i] = op(st.values[i*2+1], st.values[i*2+2])
		}
	}
}

func (st segmentTree) query(start, stop int) int {
	segments := st.segments(start, stop)
	st.propagate(segments)
	result := defaultValue
	for _, i := range segments {
		result = op(result, st.values[i])
	}
	return result
}

func main() {
	defer flush()

	N := readInt()
	a := make([]int, N)
	for i := 0; i < N; i++ {
		a[i] = readInt()
	}
	Q := readInt()

	st := newSegmentTree(N)
	st.build(a)
	for i := 0; i < Q; i++ {
		k := readInt()
		l := readInt()
		r := readInt()
		c := readInt()
		if k == 1 {
			st.apply(l-1, r, c)
		} else if k == 2 {
			println(st.query(l-1, r))
		}
	}
}

const (
	ioBufferSize = 1 * 1024 * 1024 // 1 MB
)

var stdinScanner = func() *bufio.Scanner {
	result := bufio.NewScanner(os.Stdin)
	result.Buffer(make([]byte, ioBufferSize), ioBufferSize)
	result.Split(bufio.ScanWords)
	return result
}()

func readString() string {
	stdinScanner.Scan()
	return stdinScanner.Text()
}

func readInt() int {
	result, err := strconv.Atoi(readString())
	if err != nil {
		panic(err)
	}
	return result
}

var stdoutWriter = bufio.NewWriter(os.Stdout)

func flush() {
	stdoutWriter.Flush()
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdoutWriter, args...)
}
